src/App.tsx

```diff
@@ -1,24 +1,284 @@
+  ), [todos]);
+
+  const completedTodos = useMemo(() => (
+    todos.filter(todo => todo.completed)
+  ), [todos]);
+
+  const visibleTodos = useMemo(() => {
+    const { ACTIVE, COMPLETED } = StatusFilter;
+
+    switch (selectedFilter) {
+      case ACTIVE:
+        return activeTodos;
+
+      case COMPLETED:
+        return completedTodos;
+
+      default:
+        return todos;
+    }
+  }, [todos, selectedFilter]);
+
+  const onDeleteCompletedTodos = async () => {
+    try {
+      await completedTodos.forEach(todo => {
+        onDeleteTodo(todo.id);
+      });
+    } catch {
+      setErrorMessage(ErrorMessage.REMOVE);
+    }
+  };
+
+  const onUpdateTodo = useCallback(async (
+    todoToUpdate: Todo, newTitle?: string,
+  ) => {
+    if (newTitle === todoToUpdate.title) {
+      return;
+    }
+
+    if (newTitle === '') {
+      onDeleteTodo(todoToUpdate.id);
+
+      return;
+    }
+
+    try {
+      setLoadingTodoIds((prevIds) => [...prevIds, todoToUpdate.id]);
+      const newTitleData = {
+        title: newTitle,
+      };
+
+      const newCompletedData = {
+        completed: !todoToUpdate.completed,
+      };
+
+      const updateTodoData: TodoUpdateData = newTitle
+        ? newTitleData
+        : newCompletedData;
+
+      const updatedTodo: Todo = await (
+        updateTodo(todoToUpdate.id, updateTodoData)
+      );
+
+      setTodos(prevTodos => (
+        prevTodos.map(todo => (
+          todo.id === todoToUpdate.id
+            ? updatedTodo
+            : todo
+        ))));
+
+      setLoadingTodoIds((prevIds) => (
+        prevIds.filter(prevId => prevId !== todoToUpdate.id)));
+    } catch {
+      setErrorMessage(ErrorMessage.UPDATE);
+    }
+  }, []);
+
+  const handleToggleAll = async () => {
+    try {
+      const todosToUpdate = isToggleActive
+        ? completedTodos
+        : activeTodos;
+
+      await todosToUpdate.forEach(todo => {
+        onUpdateTodo(todo);
+      });
+    } catch {
+      setErrorMessage(ErrorMessage.UPDATE);
+    }
+  };
+
   if (!USER_ID) {
     return <UserWarning />;
   }
 
   return (
-    <section className="section container">
-      <p className="title is-4">
-        Copy all you need from the prev task:
-        <br />
-        <a href="https://github.com/mate-academy/react_todo-app-add-and-delete#react-todo-app-add-and-delete">React Todo App - Add and Delete</a>
-      </p>
+    <div className="todoapp">
+      <h1 className="todoapp__title">todos</h1>
+
+      <div className="todoapp__content">
+        <header className="todoapp__header">
+          <button
+            type="button"
+            className={classNames('todoapp__toggle-all', {
+              active: isToggleActive,
+            })}
+            onClick={handleToggleAll}
+          />
+
+          <CreateNewTodo
+            setErrorMessage={setErrorMessage}
+            onAddTodo={onAddTodo}
+            isLoading={isLoading}
+          />
+
+        </header>
+
+        {todos.length > 0 && (
```

Comment:
```suggestion
        {todos.length && (
```

Reply:
In this case when the app is loading todos from server footer is visible.